consensus/istanbul/backend/announce.go

```diff
@@ -155,9 +197,76 @@ func (sb *Backend) sendIstAnnounce() error {
 	enodeUrl := enode.String()
 	view := sb.core.CurrentView()
 
-	msg := &announceMessage{Address: sb.Address(),
-		EnodeURL: enodeUrl,
-		View:     view}
+	useValidatorContract := true
+	var g core.Genesis
+	regVals, err := sb.retrieveRegisteredValidators()
+	// The validator contract may not be deployed yet.
+	// Even if it is deployed, it may not have any registered validators yet.
+	if err == errValidatorsContractNotRegistered || len(regVals) == 0 {
+		sb.logger.Trace("Can't retrieve the registered validators.  Only allowing the initial validator set to send announce messages", "err", err, "regVals", regVals)
+		valSet := sb.getValidators(block.Number().Uint64(), block.Hash())
+
+		regVals = make(map[common.Address]bool)
+		for _, val := range valSet.List() {
+			regVals[val.Address()] = true
+		}
+		useValidatorContract = false
+		genesisJSON, err := sb.db.Get(core.DBGenesisKey)
+		err = g.UnmarshalJSON(genesisJSON)
+		if err != nil {
+			log.Error("Unable to retrieve genesis from db", "err", err)
+		}
+	} else if err != nil {
+		sb.logger.Error("Error in retrieving the registered validators", "err", err)
+		return err
+	}
+
+	validatorsAddress, err := sb.regAdd.GetRegisteredAddressAtCurrentHeader(params.ValidatorsRegistryId)
+	if err != nil {
+		log.Warn("Registry address lookup failed", "err", err)
+	}
+
+	encryptedEnodeUrls := make(map[common.Address][]byte)
+	for addr := range regVals {
```

Comment:
I had a separate conversation with @kobigurk on if it's possible to create one encrypted enodeURL cipher text such that only the validators can decrypt.  The purpose is to drastically reduce the payload size of the `Announce` message.  Here is his response:

_I don't know of a direct and easy way.
Before getting into heavy crypto, one easy optimization you can make is encrypt an ephemeral decryption key for each validator rather than the whole message, and then encrypt the message once.

There's a method called broadcast encryption that I haven't used, which is almost exactly what you need. It does require a new kind of public keys though, so there's this overhead... I'll be happy to take a look at it if the trade off of adding a new kind of public key makes sense.

I'll try to think if there are non direct ways this could be done._

_To remain within ECIES encryption, for simplicity, how about that in each epoch, the proposer publishes a group key to the validators? This makes the epoch message larger, and that might be fine since it's daily. The proposer will distribute this group key encrypted under each public key, doesn't have to be in the actual block (though then we'll have to make sure it gets to all the validators).

The validators would have to verify this key. This could be done by either:
1. A ZKP that all the encrypted messages contain the same key (e.g., a Schnorr signature-like proof, called DLEQ).
2. In the message that the proposer distributes, they encrypt a fixed string, e.g., the block hash, the validators attempt to decrypt it with their group key. (edited)_ 

cc'ing @nicholasguo

Reply:
On a related note, Hotstuff BFT uses what is calls `authenticator complexity` to measure the cost of the algorithm.  

> The complexity measure we care about is authenticator complexity, which specifically is
> the sum, over all replicas i ∈ [n], of the number of authenticators received by replica i in the protocol to reach
> a consensus decision after GST. (Again, before GST, a consensus decision might not be reached at all in the worst case [27].) Here, an authenticator is either a partial signature or a signature. Authenticator complexity is a useful measure of communication complexity for several reasons. First, like bit complexity and unlike message complexity, it hides unnecessary details about the transmission topology. For example, n messages carrying one authenticator count the same as one message carrying n authenticators. Second, authenticator complexity is better suited than bit complexity for capturing costs in protocols like ours that reach consensus repeatedly, where each consensus decision (or each view proposed on the way to that consensus decision) is identified by a monotonically increasing counter. That is, because such a counter increases indefinitely, the bit complexity of a protocol that sends such a counter cannot be bounded. Third, since in practice, cryptographic operations to produce or verify digital signatures and to produce or combine partial signatures are typically the most computationally intensive operations in protocols that use them, the authenticator complexity provides insight into the computational burden of the protocol, as well.

If we are to use the same complexity measurement, then seems like the announce message will be n squared cost to the system.  If we eventually want to implement hotstuff BFT (to get to linear cost), then seems like it would be nice if we can somehow also get this message to be constant size.